include/mbedtls/build_info.h

```diff
@@ -80,6 +80,20 @@
 #include MBEDTLS_USER_CONFIG_FILE
 #endif
 
+/* PSA crypto specific configuration options
+ * - If config_psa.h reads a configuration option in preprocessor directive,
+ *   this symbol should be set before its inclusion. (e.g. MBEDTLS_MD_C)
```

Comment:
These two lines go directly against the change in this PR - even with the actual option mentioned (`MBEDTLS_MD_C`).

Specifically with this PR, `MBEDTLS_MD_C` is set after `config_psa.h` is included, but `config_psa.h` actually does read `MBEDTLS_MD_C` in a preprocessor directive.

Also, it feels like it should be possible to have a check script that finds at least some of these sorts of issues, which I think would help greatly. For instance, I can imagine an error

```
build_info.h: 109: MBEDTLS_MD_C is set after it is examined by config_psa.h line 800
```

Reply:
> MBEDTLS_MD_C is set after config_psa.h is included, but config_psa.h actually does read MBEDTLS_MD_C in a preprocessor directive.

Yes, but it's more complicated than that — see https://github.com/Mbed-TLS/mbedtls/issues/7609. In this particular case, the current order is actually correct, although this change only matters in an unimportant edge case. `config_psa.h` only reads `MBEDTLS_MD_C` when `MBEDTLS_PSA_CRYPTO_CONFIG` is disabled (and enables some hash-based key derivation functions), and in this case its job is to enable PSA crypto mechanisms for which the corresponding legacy API was requested by the user. If `MBEDTLS_MD_C` is only enabled in `build_info.h` as an indirect dependency of PKCS5, but not enabled explicitly by the user, then `config_psa.h` should enable those key derivation functions. A build with PKCS5 but not the MD module is weird (PKCS5 is mostly intended to work with encrypted RSA or ECC keys, and there's not much you can do with such keys without the MD module). So it doesn't really matter that we'd start enabling some spurious KDFs in that case.

The patch here is technically incorrect, but makes the file structure less cluttered so it makes it easier to work towards more correctness, so I'm inclined to accept it.

Reply:
> Also, it feels like it should be possible to have a check script that finds at least some of these sorts of issues

That's part of the goal of https://github.com/Mbed-TLS/mbedtls/issues/7609.